### PR TITLE
update AG metadata

### PR DESCRIPTION
**Pull request by wasade**
@josenavas, I can't test this locally. Will msg out out-of-band to coordinate

### CODE REVIEW

**Comment by josenavas** [comment]
@wasade the failures here look legit - mainly import errors though

File: knimin/handlers/barcode_util.py [hunk 1]
```diff
@@ -305,3 +312,48 @@ def post(self):
...
+                                      as_json=True)
+
+    if sc == 201:
+        return "Metadata for AG sample %s added into Qiita"
+    elif sc == 200:
+        # this should not happen as we first check whether the sample is
+        # already represented; this status code means that the sample is
+        # already present in the qiita study. there is a race condition
+        # though as we cannot put a lock on the remote resouce. So it is
+        # techinically possible that a sample metadata for this sample will
```
Comment (qiyunzhu) [inline knimin/handlers/barcode_util.py]: `techinically` => `technically`

File: knimin/handlers/barcode_util.py [hunk 2]
```diff
@@ -305,3 +312,48 @@ def post(self):
...
+    if sc != 200:
+        return "Unable to get sample IDs from Qiita"
+
+    existing = set(json_decode(response.body))
+    if barcode in existing:
+        return "Metadata for %s already exists in Qiita" % barcode
+
+    # the db.pulldown call returns formatted metadata, so we actually need to
+    # reparse it unfortunately.
+    md = pd.read_csv(md, sep='\t', dtype=str)
```
Comment (ElDeveloper) [inline knimin/handlers/barcode_util.py]: Unfortunately `dtype=str` is not sufficient to preserve string-only types, I suggest that you also add: `na_values=[], keep_default_na=False`.

File: knimin/handlers/barcode_util.py [hunk 3]
```diff
@@ -305,3 +312,48 @@ def post(self):
...
+    md.rename({'#SampleID': 'sample_name'}, inplace=True)
+    sc, response = qiita_client.patch('/api/v1/study/10317',
+                                      data=md.todict(),
```
Comment (ElDeveloper) [inline knimin/handlers/barcode_util.py]: Is this dictionary correctly created? I don't think the dataframe is indexed, so I don't know what is going to happen here. The safest choice might be to explicitly set the index.

File: knimin/handlers/barcode_util.py (hunk 2 above)
Comment (wasade) [inline knimin/handlers/barcode_util.py]: really??? damn. okay. very good to know. I'm assuming this is situations where its like "\t\t"?

Comment (ElDeveloper) [inline knimin/handlers/barcode_util.py]: That and NA, that's why keep_default_na is set to False.

**Comment by josenavas** [comment]
@wasade I think there still legit failures in here

**Pull request merged**